plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java

```diff
@@ -579,16 +585,87 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+            return false;
+        }
+
+        if (!checkArgument(call, 0, Variable.class::isInstance)
+            && checkArgument(call, 1, Constant.class::isInstance)) {
+            return false;
+        }
+
+        if (call.getArguments().size() == 3) {
+            return checkArgument(call, 2, Constant.class::isInstance);
+        }
+
+        return true;
+    }
+
+    protected static String likeToRegexp(Slice pattern, Optional<Slice> escape)
     {
-        // TODO: This can be done more efficiently by using a state machine and iterating over characters (See io.trino.type.LikeFunctions.likePattern(String, char, boolean))
-        String regexp = like.replaceAll(Pattern.quote("\\"), Matcher.quoteReplacement("\\\\")); // first, escape regexp's escape character
-        for (char c : REGEXP_RESERVED_CHARACTERS) {
-            regexp = regexp.replaceAll(Pattern.quote(String.valueOf(c)), Matcher.quoteReplacement("\\" + c));
-        }
-        return regexp
-                .replaceAll("%", ".*")
-                .replaceAll("_", ".");
+        String patternString = pattern.toStringUtf8();
+        char escapeChar = escape.map(ElasticsearchMetadata::getEscapeChar)
+                .orElse('0');
+        StringBuilder regex = new StringBuilder();
+        boolean escaped = false;
+        for (char currentChar : patternString.toCharArray()) {
+            checkEscape(!escaped || currentChar == '%' || currentChar == '_' || currentChar == escapeChar);
+            if (!escaped && (currentChar == escapeChar)) {
+                escaped = true;
+            }
+            else {
+                switch (currentChar) {
+                    case '%':
+                        regex.append(escaped ? "%" : ".*");
+                        escaped = false;
+                        break;
+                    case '_':
+                        regex.append(escaped ? "_" : ".");
+                        escaped = false;
+                        break;
+                    case '\\':
+                        regex.append("\\\\");
+                        break;
+                    default:
+                        // escape special regex characters
+                        if (REGEXP_RESERVED_CHARACTERS.contains(currentChar)) {
+                            regex.append('\\');
+                        }
+
+                        regex.append(currentChar);
+                        escaped = false;
+                }
+            }
+        }
+
+        checkEscape(!escaped);
+        return regex.toString();
+    }
+
+    private static void checkEscape(boolean condition)
+    {
+        if (!condition) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Escape character must be followed by '%%', '_' or the escape character itself");
+        }
+    }
+
+    private static char getEscapeChar(Slice escape)
+    {
+        String escapeString = escape.toStringUtf8();
+        if (escapeString.isEmpty()) {
+            return '0';
```

Comment:
Is it valid for escape to be empty?
Does `'0'` (`(char) 48`) have any special meaning?

Reply:
It does not.

Reply:
Oh snap :) Fixed!

Reply:
has this been fixed?

Reply:
It is now. Sorry - I've got confused :)